synapse/rest/client/v1/login.py

```diff
@@ -373,6 +375,24 @@ def register(self, http_server: HttpServer) -> None:
     async def on_GET(
         self, request: SynapseRequest, idp_id: Optional[str] = None
     ) -> None:
+        if not self._public_baseurl:
+            raise SynapseError(400, "SSO requires a valid public_baseurl")
+
+        # if this isn't the expected hostname, redirect to the right one, so that we
+        # get our cookies back.
+        requested_uri = get_request_uri(request)
+        baseurl_bytes = self._public_baseurl.encode("utf-8")
+        if not requested_uri.startswith(baseurl_bytes):
+            i = requested_uri.index(b"/_matrix")
+            new_uri = baseurl_bytes[:-1] + requested_uri[i:]
```

Comment:
It seems like it would be simpler to do `baseurl_bytes + request.uri`. Actually -- it seems odd that we rebuild the entire uri when we really only want to compare the scheme + host (+ port)?

Reply:
the idea is that we'll behave correctly here if `/_matrix` isn't at the root of the path. (Of course, there are plenty of other places that won't work...)

Reply:
So this should redirect from https://foo.bar/whatever/_matrix/... -> https://public.baseurl/_matrix/...?

Reply:
correct. I guess I should spell that out in a comment.